File: python/leetcode/298.py
class Solution:
    def longestConsecutive(self, root):
        if not root: return 0
        return self.help(root, 0, root.val - 1)

    def help(self, root, length, preVal):
        if not root: return length
        cur = 1 if preval + 1 != root.val else length + 1
        return max(cur, max(self.help(root.left, cur, root.val), self.help(root.right, cur, root.val)))

File: cpp/leetcode/298.cpp
class Solution {
public:
    int longestConsecutive(TreeNode* root) {
        TreeNode* pre = NULL;
        int res = 0;
        return dfs(root, pre, res);
    }
    
    int dfs(TreeNode* root, TreeNode* pre, int res){
        if(!root) return res;
        res = (pre && pre->val + 1 == root->val)? res + 1 : 1;
        return max(res, max(dfs(root->left, root, res), dfs(root->right, root, res)));
    }
};